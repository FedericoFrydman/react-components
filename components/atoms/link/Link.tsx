import { forwardRef, Ref } from 'react';
import cn from 'classnames';

import { Icon, IconNames } from '../icon/Icon';
import { Text, TextSize } from '../text/Text';

import './link.scss';

export type LinkTarget = '_blank' | '_self' | '_parent' | '_top';

export type LinkPosition = 'left' | 'right';

export type LinkWeight = 'normal' | 'semibold';

export interface LinkProps extends Omit<React.HTMLProps<HTMLAnchorElement>, 'target' | 'css'> {
  /**
   * Remove the default margin below the text.
   */
  noMargin?: boolean;
  /**
   * Specifies where to open the link
   */
  target?: LinkTarget;
  /**
   * The text to be rendered
   */
  text: string;
  /**
   * Size of the text.
   * @default `medium` paramater
   */
  textSize?: TextSize;
  /**
   * The font weight.
   * @default `normal` paramater
   */
  textWeight?: LinkWeight;
  /**
   * Display an icon in addition to the text to help to identify the action. The color will be same of the text
   */
  icon?: IconNames;
  /**
   * Icon at left/right of the link. Default right.
   */
  iconPosition?: LinkPosition;
  /**
   * Disables link text decoration (i.e. the underline under the link).
   * @deprecated Default is no decoration. Will be removed in future verison.
   */
  noDecoration?: boolean;
  /**
   * Show link text decoration underline always.
   */
  decoration?: boolean;
  /**
   * Custom classes for the component.
   * @default false paramater
   */
  classNames?: cn.Argument;
  /**
   * The text to be rendered, unless you set the text prop.
   */
  children?: string | number;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLAnchorElement>;
  /**
   * An event id that will be used by analytics and test automation.
   */
  pplsiEventId?: string;
}

const LinkComponent = (
  {
    href = '',
    target,
    noMargin,
    text,
    icon,
    children,
    iconPosition = 'right',
    noDecoration,
    decoration = false,
    classNames = [],
    pplsiEventId,
    textWeight = 'normal',
    textSize = 'medium',
    ...props
  }: LinkProps,
  ref: LinkProps['ref'],
) => {
  const classLink = 'lsux-link';
  const cnLink = cn(
    {
      [`${classLink}`]: true,
      [`${classLink}--no-margin`]: noMargin,
      [`${classLink}--decoration`]: decoration,
      [`${classLink}--no-decoration`]: noDecoration,
    },
    classNames,
  );

  return (
    <a {...props} ref={ref} href={href} data-pplsi-event-id={pplsiEventId} className={cnLink} target={target}>
      {icon && iconPosition === 'left' && <Icon name={icon} size="small" classNames={['mr-2']} />}
      {text ? <Text text={text} textWeight={textWeight} textSize={textSize} /> : children}
      {icon && iconPosition === 'right' && <Icon name={icon} size="small" classNames={['ml-2']} />}
    </a>
  );
};

export const Link = forwardRef<HTMLAnchorElement, LinkProps>(LinkComponent);
