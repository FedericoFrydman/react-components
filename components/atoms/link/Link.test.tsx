import { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import { shallow, ShallowWrapper } from 'enzyme';
import cn from 'classnames';

import { Link, LinkProps } from './Link';

describe('Link Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;
  let pplsiEventId: string;

  beforeEach(() => {
    wrapper = shallow(<Link text="button" />);
    testId = 'test link';
    pplsiEventId = 'test prop';
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the text the provided classNames', () => {
          const text = renderLink({ classNames });

          expect(text).toHaveClass(cn(classNames), { exact: false });
        });
      });

      describe('when decoration is true', () => {
        const className = 'lsux-link--decoration';

        it('gives the text the provided classNames', () => {
          const text = renderLink({ decoration: true });

          expect(text).toHaveClass(cn(className), { exact: false });
        });
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLAnchorElement>();

    it('provides a ref to the spinner', () => {
      expect(renderLink({ ref })).toEqual(ref.current);
    });
  });

  const renderLink = (props: Omit<LinkProps, 'text'>): HTMLAnchorElement => {
    render(<Link {...props} text="Button" data-testid={testId} />);

    return screen.getByTestId(testId);
  };

  describe('unspecified html attributes', () => {
    it('allows pplsi event id data attribute to be passed to the underlying link element', () => {
      render(<Link pplsiEventId={pplsiEventId} data-testid={testId} text="button" />);

      const link = screen.getByTestId(testId);

      expect(link).toHaveAttribute('data-pplsi-event-id', pplsiEventId);
    });
  });
});
