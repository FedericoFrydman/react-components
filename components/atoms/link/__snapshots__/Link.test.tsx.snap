// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link Component should render the component 1`] = `
<a
  className="lsux-link"
  href=""
>
  <ForwardRef(TextComponent)
    text="button"
    textSize="medium"
    textWeight="normal"
  />
</a>
`;
